pkg/extsvc/bitbucketserver/client.go

```diff
@@ -133,9 +135,13 @@ func (c *Client) do(ctx context.Context, req *http.Request, result interface{})
 		nethttp.ClientTrace(false))
 	defer ht.Finish()
 
+	startWait := time.Now()
 	if err := c.RateLimit.Wait(ctx); err != nil {
 		return err
 	}
+	if d := time.Since(startWait); d > 200*time.Millisecond {
+		log15.Warn("Bitbucket self-enforced API rate limit: request delayed longer than expected due to rate limit", "delay", d)
```

Comment:
The self-enforced wording makes it unclear to me who is enforcing the rate limit (Sourcegraph or Bitbucket). Can you clarify?

Reply:
We use this wording in other locations, not just here. It's also consistent with how I've communicated this with multiple users.

If Bitbucket enforced the rate limit, I think it is fair to say we would just say `Bitbucket API rate limit` and omit `self-enforced` entirely. The log message also comes from Sourcegraph, which is another hint that "self" is Sourcegraph.

- We could say `Bitbucket Sourcegraph-enforced API rate limit` or something but IMO that is more confusing.
- We could say `Bitbucket API rate limit enforced by Sourcegraph` but that seems to be more of a mouthful, and also leaves ambiguity about "Is this just saying Sourcegraph respects the rate limit Bitbucket has?".

I don't know of better wording here. Suggestions?

Reply:
> We use this wording in other locations, not just here

Examples? A quick search didn't yield any:
https://sourcegraph.com/search?q=r:sourcegraph/+self%20enforced

> I don't know of better wording here. Suggestions?

It is still a bit unclear to me what the exact semantics are here. Looking at the code it seems like we are honoring Bitbucket's configured rate limit. If so, I would suggest this wording:

```suggestion
		log15.Warn("Bitbucket is rate limiting API requests", "delay", d)
```

Is Warn the right level here? Will customers that do have a rate limit configured see log spam?

Reply:
You are missing the dash in your query: https://sourcegraph.com/search?q=r:sourcegraph/+self-enforced

> It is still a bit unclear to me what the exact semantics are here. Looking at the code it seems like we are honoring Bitbucket's configured rate limit.

Bitbucket Server has no rate limiting, it simply does not exist. Because of this, we impose our own rate limiting on our usage of Bitbucket Server's API. That is to say, _it is self-enforced_, not enforced by Bitbucket.

We choose the limits, therefor there is nothing e.g. that can be configured on Bitbucket Server to give Sourcegraph more of an API quota than what it is already using.

Reply:
> Is Warn the right level here? Will customers that do have a rate limit configured see log spam?

If customers see this warning, it indicates a bug on our end (using too many code host API requests), so yes it is the right level.